[Drake] Vending step5

DrakeYang opened the pull request:
피드백 부탁드립니다!

DrakeYang committed:
DrinkType decoding 방식 수정 : rawvalue 를 받아서 enum 으로 변경

DrakeYang committed:
기본음료 추가함수 삭제 : 모든 음료는 0개부터 시작

DrakeYang committed:
음료 추가시 notification 을 보내도록 추가

DrakeYang committed:
음료변화 noti 를 받아서 실행할 함수 추가 : drinkCountChanged .
음료추가 noti 가 발생하면 받아서 실행함수에게 noti를 보내는 옵저버 추가

DrakeYang committed:
진행된 내용을 README 에 추가

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -120,28 +131,38 @@ class ViewController: UIViewController {
     }
+    
+    // notification 생성
+    
```
주석내용만 있고 코드가 없네요?

DrakeYang replied:
원래 아무것도 없는 부분 맞습니다 ; 주석 정리하다가 밑으로 밀려간것 같네요

godrm replied:
주석이 코드와 일치하지 않는것도 주의해주세요.

DrakeYang replied:
넵

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -96,11 +101,17 @@ class ViewController: UIViewController {
     }
```
drinkCounts 가 `[UILabel]` 이었군요. 화면 요소는 `-Label` 처럼 화면 요소 이름을 붙여주는 게 좋습니다.

DrakeYang replied:
변수명을 drinkCounts-Label 로 하라는 말씀이신가요?

godrm replied:
`-Label` 에 `-` 는 앞부분에 뭐가 있고 접미사라는 표현이죠 ㅋㅋ
Label을 뒷부분 어딘가에 포함시켜보세요.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -120,28 +131,38 @@ class ViewController: UIViewController {
+        
         self.drink01View.image = UIImage(named: "Drink01.jpg")
+        
         setBorderRadius()
```
이 부분은 뭐죠? 
왜 여기만 drink01View 가 하나 별도로 있는건가요?

DrakeYang replied:
스토리보드 와 코드 두가지 모두 써보려고 했습니다.
나머지 드링크는 스토리보드에서 넣었구요

godrm replied:
아 그런거면 오케이인데. 차라리 뷰를 개선할 때 한 번씩 바꿔보는 게 일관성이 있을 것 같기도 합니다

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -120,28 +131,38 @@ class ViewController: UIViewController {
+        // 노티를 보는 옵저버. 노티가 발생하면 해당 함수를 실행한다
+        NotificationCenter.default.addObserver(self, selector: #selector(self.drinkCountChanged(notification:)), name: .drinkCountChanged , object: nil)
+        
         super.viewDidLoad()
```
super.viewDidLoad()를 호출할꺼면 가장 먼저 하는게 좋습니다.

DrakeYang replied:
옙

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -96,11 +101,17 @@ class ViewController: UIViewController {
     }
```
drinkCount 가 UILabel 이라면 superview 에서 viewWithTag() 로 찾는 방법이 있습니다. 
이렇게 목록에서 비교할 필요는 없습니다.

godrm replied:
싱글톤이면 appDelegate.sharedVendingMachine 로 접근할 필요가 없습니다.
VendingMachine.shared() 로 직접 접근하는 편이 낫습니다.
하지만 결과적으로는 VendingMachine 객체에 의존하는 것이기 때문에 AppDelegate에서 ViewController에 의존성을 주입해주는 게 좋습니다. 이 부분을 한 번 개선해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -40,8 +47,6 @@ class ViewController: UIViewController {
             case 6 : try! appDelegate.sharedVendingMachine.addBasicDrink(drinkType: .energyDrink)
             default : throw OutputView.errorMessage.wrongDrink
             }
```
switch-case 로 비교하지 않고 매핑할 수 있는 방법을 한 번 고민해보세요.

DrakeYang committed:
super.viewDidLoad 를 제일 먼저 실행하도록 순서 변경

DrakeYang committed:
음료추가 함수의 입력인자 변경 : 음료타입 > 정수

DrakeYang committed:
음료추가함수 에서 음료타입 생성방법 변경 : init > rawValue

DrakeYang committed:
음료추가버튼 로직 변경 : sender.tag 를 switch > tag 바로 사용

DrakeYang committed:
음료정보>태그번호 변환 함수 로직 변경 : switch 구문 삭제. DrType.rawValue 를 바로 리턴

DrakeYang committed:
음료타입.rawValue 순서 변경 : 스토리보드와 맞춤

DrakeYang committed:
음료타입 요구하는 함수들 로직변경 : rawValue 로 바로 탐색.
음료카운트 배열에서 반복문 제거.
주석 없는 부분엔 주석 추가, 불필요한 주석 삭제

DrakeYang committed:
음료사진 출력방법 스토리보드로 통일 - 기존 사진로딩코드 삭제..
음료재고 갱신 함수에서 초기화 부분 삭제. 초기화는 viewDidLoad 에서 한번만.
음료재고 갱신 방법 수정 : [label] 반복문 > withTag 함수 사용

DrakeYang committed:
lable 배열들 변수명 변경 : -Label 추가

DrakeYang committed:
자판기 객체용 옵셔널 변수 추가.
사용하지 않는 디코딩용 init 제거

DrakeYang committed:
뷰컨트롤러에 자판기 객체 의존성 주입

DrakeYang commented:
피드백 부탁드립니다.
